internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -50,71 +54,60 @@ func (k *Kubeadm) PreFunc(data *userdata.UserData) (err error) {
 		return err
 	}
 
-	if data.Services.Kubeadm.IsBootstrap() {
-		// Write out all certs we've been provided
-		certs := []struct {
-			Cert     *x509.PEMEncodedCertificateAndKey
-			CertPath string
-			KeyPath  string
-		}{
-			{
-				Cert:     data.Security.Kubernetes.CA,
-				CertPath: constants.KubeadmCACert,
-				KeyPath:  constants.KubeadmCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.SA,
-				CertPath: constants.KubeadmSACert,
-				KeyPath:  constants.KubeadmSAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.FrontProxy,
-				CertPath: constants.KubeadmFrontProxyCACert,
-				KeyPath:  constants.KubeadmFrontProxyCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.Etcd,
-				CertPath: constants.KubeadmEtcdCACert,
-				KeyPath:  constants.KubeadmEtcdCAKey,
-			},
-		}
-
-		for _, cert := range certs {
-			if cert.Cert == nil {
-				continue
-			}
-			if err = writeKubeadmPKIFiles(cert.Cert, cert.CertPath, cert.KeyPath); err != nil {
-				return err
-			}
-		}
-	} else if data.Services.Kubeadm.IsControlPlane() {
-		if data.Services.Trustd == nil || data.Services.Trustd.BootstrapNode == "" {
-			return nil
-		}
+	// Write out all certs we've been provided
+	if err = writeKubeadmPKIFiles(data); err != nil {
+		return err
+	}
 
-		creds, err := basic.NewCredentials(data.Services.Trustd)
-		if err != nil {
+	// Run kubeadm init phase certs all. This should fill in whatever gaps
+	// we have in the provided certs.
+	if data.Services.Kubeadm.IsBootstrap() {
+		cmd := exec.Command(
+			"kubeadm",
+			"init",
+			"phase",
+			"certs",
+			"all",
+			"--config",
+			constants.KubeadmConfig)
+		if err = cmd.Run(); err != nil {
 			return err
 		}
+	}
 
-		files := []string{
-			constants.AuditPolicyPathInitramfs,
-			constants.EncryptionConfigInitramfsPath,
-		}
+	if data.Services.Kubeadm.IsWorker() || data.Services.Kubeadm.IsBootstrap() || data.Services.Trustd == nil {
+		log.Println("Skipping retrieval of files from peers via trustd")
+		return nil
+	}
 
-		conn, err := basic.NewConnection(data.Services.Trustd.BootstrapNode, constants.TrustdPort, creds)
-		if err != nil {
-			return err
-		}
+	// Get a list of files we need to request
+	// ( filtered by ones we already have )
+	filelist := fileSet()
 
-		client := proto.NewTrustdClient(conn)
+	// Initialize trustd peer client connection
+	var trustds []proto.TrustdClient
+	if trustds, err = createTrustdClients(data); err != nil {
+		return err
+	}
 
-		if err := writeFiles(client, files); err != nil {
-			return err
+	// Get assets from remote nodes
+	for _, fileRequest := range filelist {
+		// Have a single chan shared across all clients
+		// for a given file
+		content := make(chan []byte, len(trustds))
+		donezo := make(chan struct{})
+
+		// kick off a goroutine for each trustd client
+		// to fetch the given file
+		for _, trustdClient := range trustds {
+			go download(trustdClient, fileRequest, content, donezo)
```

Comment:
I think we need to get one instance of data for each file, does it make sense to ask for all the files each trustd endpoint?

I mean we don't need to launch goroutine per file * trustd endpoint, but goroutine per trustd endpoint, each with the same list of files, and after that we need some kind of 'reduce' phase when we remove duplicates and give errors if some files can't be fetched? 

probably I'm overcomplicating things, but at the same time it feels like `writeTrustdFiles` should run when there's a single instance of file downloaded (which is enough?)

Reply:
I was messing with something like this yesterday and ended up have `download` return `chan <- []byte`. Basically in `download` the first client to get the content wins, and then `writeTrustdFiles` accepts `content []byte` parameter.

Reply:
I wanted to set up a model for a fan in approach ( think I got the naming correct on this ). We reach out to all trustd instances and try to fetch the necessary files and mux them afterwards. This approach seemed like it painted the most reliable picture so that we could not be stuck relying on a single node for the information but get that information from any control plane node ( since these resources should be identical ). And by doing a goroutine per file per trustd instance, I was hoping it'd be the quickest/most performant ( in that the max time should only be the response time for a single file versus a list of files ). 
If there's opposition to this approach ( I know Andrew and I have gone back and forth on it ), I can look at other implementations.

Reply:
I'm a fan of that approach 100%. For me its the implementation details. You can fan out and try all trustd clients in `download`, and then fan in and return one instance of `content` that gets passed into `writeTrustdFiles`.

Reply:
I have just one concern here:

* we can do fan out, retries, whatever in `download`, but I think it should be sequential (without channel, concurrent goroutines) from `download` to `write`.